### ISSUE TITLE
Printer type?

### ISSUE THREAD

**Issue by Kulitorum** [slaytonrnd/CartridgeWriter#4]
Do you know what printer type a SST 768 is? 

I don't understand the Fox, Fox2 etc. How do they refer to models?

**Comment by slaytonrnd** [comment]
I'm not sure how all of the machines map either.  For the Dimension 1200es I was working with I used Prodigy and I recently found out that a Fortus 400c is a Fox type machine.  No clue on the SST 768.  Whenever I've seen anyone ask Benjamin he suggests just trying them all until you get meaningful output.  If you select the wrong one and press Read CartridgeWriter will just throw an exception and bomb right now.  Not much in the way of error handling just yet...

**Comment by Kulitorum** [comment]
Ok. Trying random stuff just freezes the program for me, so I guess I should look at the arduino code for a while. Thank you very much for all your help.

**Comment by slaytonrnd** [comment]
It will freeze if it is waiting on the Arduino for a response (like if the EEPROM is not hooked up properly or not hooked up at all).  Check the Rx/Tx lights and see if there is any activity and see where it gets stuck.  Also I had to add a pause to the program (Thread.Sleep(10)) to wait for a response from the Micro.  I'm guessing that the timing should not change, but something to keep an eye out for depending on where it is getting stuck.  Good luck!

**Comment by Kulitorum** [comment]
Hmm, I can't make it work - just freezes whatever I do. Do you power your arduino externaly?

**Comment by Kulitorum** [comment]
Ok, I made it work by changing he WaitForChip routine to always start with waiting for 3 seconds.

```
    private void WaitForChip(SerialPort sp)
    {
        Thread.Sleep(3000);
        while (!PollForChip(sp))
            Thread.Sleep(500);

        sp.DiscardInBuffer();
    }
```

**Comment by Kulitorum** [comment]
Oh, I also changed the USB speed to 115200 - don't know if that matters.

**Comment by slaytonrnd** [comment]
Congratulations on getting it working and thanks for letting me know your results!  Looking at the code I can see where the serial device might not have enough time to recover between the "sp.DiscardInBuffer()" call and the "WaitForChip(sp)" call.  Did you happen to try a smaller value for Thread.Sleep before settling on 3000?  I'd like to add a pause in the code, but want to make it as small as possible.  I am also going to open some enhancements to make troubleshooting easier.

**Comment by Kulitorum** [comment]
Well, I tried 1000ms, that did not work for me. I'd rather have it a bit slow but stable, then fast and unstable, so I settled for the 3sek.

**Comment by slaytonrnd** [comment]
Totally makes sense.  I went ahead and added a configuration file (CartridgeWriter.exe.config) with default values for BaudRate = 9600 and SerialInitializationWait = 2000, but you can change them to whatever you want.  I figured other people may also run into similar issues and need custom values so might as well make it user settable.  I'm currently updating the readme with notes about the values.  I don't have a way to test this, but I kind of figured that maybe the problem was how long it was taking to open the serial device so I've also changed when the wait happens.  There used to be a call to sp.DiscardInBuffer() before the wait, but now it happens after the wait and the code looks like so:

```
        private void WaitForChip(SerialPort sp)
        {
            Thread.Sleep(serialInitializationWait);

            sp.DiscardInBuffer();

            while (!PollForChip(sp))
                Thread.Sleep(500);

            sp.DiscardInBuffer();
        }
```

This works fine with the Micro; when you get a chance would you be able to check with your Mega and see if it works for you?  Thanks!

**Comment by Bartimaeus-** [comment]
An SST 768 is known by the code name "schooner". What would it take to add support for another printer type? I'm using a BST 1200 ("oahu"). I have dumps of the EEPROMS if that helps.

**Comment by Kulitorum** [comment]
CartridgeWriter already supports the BST 1200. It is compatable with the Prodigy, just like the 768.

**Comment by Bartimaeus-** [comment]
You're right, sorry. I realized my mistake last night after some reading. If I could ask another uneducated question, would it be difficult to add support for the bus pirate? bvanheu has some python scripts for reading and writing using the bus pirate , and it would be awesome to have a gui to use.

**Comment by slaytonrnd** [comment]
I took a look at bvahheu's bus pirate scripts and it looks like it would not be too hard to port the code over, but I don't have a bus pirate so it would probably need a few rounds of back and forth testing / coding to make sure it works right.  I'm pretty booked right now, but might have some time to look into it further on Sunday (2/22).  Feel free to add a new issue to add Bus Pirate support  to CartridgeWriter if you want to go that route and I'll get to it as I can.
